### PR TITLE
[Client] Created Parent Class for PageViews, PageDataHandler, ListDataHandler.

### PR DESCRIPTION
**Pull request by Lorraine-Lyu**
This branch isn't basing off from the refactored version and collegedatahandler is not a class.
We can delete collegePageView and scholarshipPageView; collegeListView and ScholarshipListView later merging.
For now I just keep using the central controller and all sub classes.

### CODE REVIEW

File: src/main/webapp/js/finscholar/collegepageview/collegepageview.js [hunk 1]
```diff
@@ -16,15 +16,18 @@
...
+const {SinglePageView} = goog.require('finscholar.singlepageview');
 const {loadCollegeData} = goog.require('datahandlers.collegepage');
 
 /** Class for the college page view. */
-class CollegePageView {
-  constructor() {}
+class CollegePageView extends SinglePageView {
+
+  constructor() {
+    super(undefined, undefined);
```
Comment (Lorraine-Lyu) [inline src/main/webapp/js/finscholar/collegepageview/collegepageview.js]: currently CollegeDataHandler is not a class and it directly renders college page to view. So the call to parent here is just a place holder. We'll think about overall structure of our code more during alpha.

File: src/main/webapp/js/finscholar/singlepageview.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+   * @constructor
+   */
+  constructor(dataHandler, template) {
+    /** 
+     * @private @const {!ScholarshipDataHandler | !CollegePageDataHandler} dataHandler_ 
+     * The object fetches and formats scholarship data.
+     */
+    this.dataHandler_ = dataHandler;
+    /**
+     * @private @const {!function(?):Element}
```
Comment (Lorraine-Lyu) [inline src/main/webapp/js/finscholar/singlepageview.js]: The input of the function can be either a scholarship object or a college pbject.

File: src/main/webapp/js/finscholar/collegepageview/collegepageview.js (hunk 1 above)
Comment (MikolJiron) [reply src/main/webapp/js/finscholar/collegepageview/collegepageview.js]: I'll be working off this branch today, should I make CollegeDataHandler into a class?

**Review by MikolJiron — APPROVED** [review]

**Review by MikolJiron — APPROVED** [review]

**Review by CarlosM10011 — APPROVED** [review]

File: src/main/webapp/js/finscholar/collegepageview/collegepageview.js [hunk 3]
```diff
@@ -16,15 +16,18 @@
...
+  }
 
   /** Render the college page. */
-  async renderView(element) {
+  async renderView(element, id) {
```
Comment (tinagrasso) [inline src/main/webapp/js/finscholar/collegepageview/collegepageview.js:30]: id is never used?

File: src/main/webapp/js/finscholar/scholarshippageview/scholarshippageview.js [hunk 4]
```diff
@@ -16,50 +16,21 @@
...
-      throw new Error(`Cannot get data for scholarship ${id}, message: ${e}`);
-    }
-    try {
-      container.innerHTML = scholarshippage({scholarship: scholarshipData});      
-    } catch(e) {
-      console.log(e);
-      // Throws the error to the caller, and the caller will render an error page instead.
-      throw new Error(`Failed to generate html: ${e}`);
-    }
+    super(new ScholarshipDataHandler(), scholarshippage);
```
Comment (tinagrasso) [inline src/main/webapp/js/finscholar/scholarshippageview/scholarshippageview.js:31]: Same comment as the other PR. Why move this code out?

File: src/main/webapp/js/finscholar/singlepageview.js (hunk 2 above)
Comment (tinagrasso) [reply src/main/webapp/js/finscholar/singlepageview.js]: Same comment as other PR. Don't allow multiple types. Use inheritance to guarantee the structure of the class.

**Review by tinagrasso — CHANGES_REQUESTED** [review]

File: src/main/webapp/js/finscholar/collegepageview/collegepageview.js (hunk 3 above)
Comment (Lorraine-Lyu) [reply src/main/webapp/js/finscholar/collegepageview/collegepageview.js:30]: Oh the loadCollegeData method doesn't query by id in this version. This collgepageview class is just a framework for #74 and #73

File: src/main/webapp/js/finscholar/collegepageview/collegepageview.js [hunk 5]
```diff
@@ -16,15 +16,19 @@
...
 goog.module('finscholar.collegepageview');
 
-const GoogDom = goog.require('goog.dom');
+const {SinglePageView} = goog.require('finscholar.singlepageview');
 const {loadCollegeData} = goog.require('datahandlers.collegepage');
 
 /** Class for the college page view. */
-class CollegePageView {
-  constructor() {}
+class CollegePageView extends SinglePageView {
```
Comment (Lorraine-Lyu) [inline src/main/webapp/js/finscholar/collegepageview/collegepageview.js:23]: This class is just an empty template which renders hard coded data from server. The actual implementation is in #74 #73

File: src/main/webapp/js/finscholar/singlepageview.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+ * @public
+ */
+class SinglePageView {
+
+  /**
+   * @constructor
+   */
+  constructor(template) {
+    /** 
+     * @private @const {?PageDataHandler} dataHandler_
```
Comment (Lorraine-Lyu) [inline src/main/webapp/js/finscholar/singlepageview.js]: Will add PageDataHandler and PageObject definition later in other branches.

File: src/main/webapp/js/finscholar/singlepageview.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    /**
+     * @private @const {!function(PageObject):Element}
+     */
+    this.template_ = template;
+  }
+
+  /** @param {!PageDataHandler} dataHandler */
+  setDataHandler(dataHandler) {
+    this.dataHandler_ = dataHandler;
+  }
```
Comment (tinagrasso) [inline src/main/webapp/js/finscholar/singlepageview.js]: Why is this needed? You have it as a const above. This probably prevents your code from compiling.

File: src/main/webapp/js/finscholar/scholarshippageview/scholarshippageview.js [hunk 8]
```diff
@@ -16,50 +16,25 @@
...
-      throw new Error(`Cannot get data for scholarship ${id}, message: ${e}`);
-    }
-    try {
-      container.innerHTML = scholarshippage({scholarship: scholarshipData});      
-    } catch(e) {
-      console.log(e);
-      // Throws the error to the caller, and the caller will render an error page instead.
-      throw new Error(`Failed to generate html: ${e}`);
-    }
+    this.dataHandler = new ScholarshipDataHandler();
```
Comment (tinagrasso) [inline src/main/webapp/js/finscholar/scholarshippageview/scholarshippageview.js]: Don't use a setter. Pass this into the constructor if you want to keep it as a separate class.

**Review by tinagrasso — CHANGES_REQUESTED** [review]

File: src/main/webapp/js/finscholar/scholarshippageview/scholarshippageview.js (hunk 4 above)
Comment (Lorraine-Lyu) [reply src/main/webapp/js/finscholar/scholarshippageview/scholarshippageview.js:31]: We'll use a completely different controller after refactoring the code (it's split into multiple branches now); at that time we won' need this class.

File: src/main/webapp/js/datahandlers/listdatahandler.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+/** 
+ * This class loads scholarship/college data from the backend and formats it for soy templates.  
+ */
+class ListDataHandler {
+
+  /** 
+   * @returns {Promise<number>} 
+   * The total number of scholarship stored in backend. 
+   */
+  async getTotalNumber() {}
```
Comment (tinagrasso) [inline src/main/webapp/js/datahandlers/listdatahandler.js]: Why not implement this method? Isn't it just returning a constant?

File: src/main/webapp/js/datahandlers/scholarshipdatahandler.js [hunk 10]
```diff
@@ -81,7 +84,7 @@ class ScholarshipDataHandler {
    * @param {string} id The uuid of the scholarship data.
    * @return The formatted scholarship JS object map.
    */
-  async fetchAndFormatSingleScholarshipData(id) {
+  async fetchAndFormatData(id) {
```
Comment (tinagrasso) [inline src/main/webapp/js/datahandlers/scholarshipdatahandler.js]: Move to base class.

File: src/main/webapp/js/datahandlers/scholarshipdatahandler.js [hunk 11]
```diff
@@ -81,7 +84,7 @@ class ScholarshipDataHandler {
...
     let data = undefined;
     try {
       data = await this.fetchScholarshipJson_(id);
```
Comment (tinagrasso) [inline src/main/webapp/js/datahandlers/scholarshipdatahandler.js]: Rename to something generic and move to base class.

```
async fetchData(path, parameters) { ... }
```

File: src/main/webapp/js/datahandlers/scholarshipdatahandler.js [hunk 12]
```diff
@@ -16,6 +16,7 @@
 
 goog.module('datahandlers.scholarshipdatahandler');
 const {Map: SoyMap} = goog.require('soy.map');
+const {SinglePageDataHandler} = goog.require('datahandlers.singlepagedatahandler');
 const {addSpaceToCamelCase} = goog.require('datahandlers.utils');
 
 const NA = 'N/A';
```
Comment (tinagrasso) [inline src/main/webapp/js/datahandlers/scholarshipdatahandler.js:22]: Make a map of fields to display names. This way you don't need to do any string manipulations, especially if you're already hardcoding this.

```
const FIELDS = [
  { fieldName:  'academicRequirements', displayName: 'Academic Requirements' },
];
```

File: src/main/webapp/js/datahandlers/scholarshipdatahandler.js [hunk 13]
```diff
@@ -59,20 +60,22 @@ class ScholarshipDataHandler {
     }
```
Comment (tinagrasso) [inline src/main/webapp/js/datahandlers/scholarshipdatahandler.js:61]: What is this `key` data used for?

I'm really confused about what the requirementsMap is doing. REQUIREMENTS isn't a map so there will never be a value. Also you're only iterating through fields in REQUIREMENTS so you'll never hit the else block.

File: src/main/webapp/js/datahandlers/scholarshiplistdatahandler.js [hunk 14]
```diff
@@ -62,7 +62,7 @@ class ScholarshipListDataHandler {
     try {
       let responseList = await fetch(SCHOLARSHIP_LIST_ENDPOINT + queryString);
       scholarshipList = await responseList.json();
-      scholarshipList = scholarshipList.map(e => this.formatListItem(e));
+      scholarshipList = scholarshipList.map(e => this.formatListItem_(e));
```
Comment (tinagrasso) [inline src/main/webapp/js/datahandlers/scholarshiplistdatahandler.js]: Don't reassign variables to a different type.

File: src/main/webapp/js/datahandlers/scholarshiplistdatahandler.js [hunk 15]
```diff
@@ -62,7 +62,7 @@ class ScholarshipListDataHandler {
...
     } catch(e) {
       console.log(e);
       throw e;
```
Comment (tinagrasso) [inline src/main/webapp/js/datahandlers/scholarshiplistdatahandler.js]: You can make this method way more generic. `type` can be an instance variable.

File: src/main/webapp/js/datahandlers/scholarshiplistdatahandler.js [hunk 16]
```diff
@@ -76,8 +76,9 @@ class ScholarshipListDataHandler {
   /**
    * @param {Object} item The schoarship object.
    * @returns {!Array<!Array<string>>} The formatted scholarship list item.
+   * @private
    */
-  formatListItem(item) {
+  formatListItem_(item) {
     const amount = (item[AMOUNT_PER_YEAR] || UNKNOWN);
```
Comment (tinagrasso) [inline src/main/webapp/js/datahandlers/scholarshiplistdatahandler.js]: Are you shortening a number here with an ellipsis? I think it's best to display the whole number.

**Review by tinagrasso — CHANGES_REQUESTED** [review]
Took a first pass through the datahandler stuff, but this PR is a little overwhelming with how big it is. Let's get through the datahandler stuff and then I'll com back to the listview stuff :)

**Review by hzimz — APPROVED** [review]

File: src/main/webapp/js/datahandlers/scholarshipdatahandler.js (hunk 12 above)
Comment (Lorraine-Lyu) [reply src/main/webapp/js/datahandlers/scholarshipdatahandler.js:22]: I added a map

**Review by tinagrasso — APPROVED** [review]
